x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/support/ApiKeyGenerator.java

```diff
@@ -43,6 +50,14 @@ public void generateApiKey(Authentication authentication, CreateApiKeyRequest re
             return;
         }
         apiKeyService.ensureEnabled();
+        if (request.getMetadata() != null && false == request.getMetadata().isEmpty()) {
+            final Version smallestNonClientNodeVersion = clusterService.state().nodes().getSmallestNonClientNodeVersion();
+            if (smallestNonClientNodeVersion.before(FLATTENED_FIELD_TYPE_INTRODUCED)) {
+                listener.onFailure(new IllegalArgumentException("API metadata requires all nodes to be at least ["
+                    + FLATTENED_FIELD_TYPE_INTRODUCED + "], got [" + smallestNonClientNodeVersion + "]"));
+                return;
+            }
+        }
```

Comment:
I don't think this belongs in the generator. 
It's effectively the same class of checks as whether the index exists, is up to date, etc. They belong in the service.

Reply:
I was trying to fail earlier without having to go through retrieving roles. But it is better to check this inside ApiKeyService for maximum possible coverage as well.